### ISSUE TITLE
Failed to register background callback

### ISSUE THREAD

**Issue by aljkor** [ABausG/home_widget#27]
Hi,

when I try to register a background callback to register clicks on a home widget, I get a ClassCastException in HomeWidgetPlugin.kt: 

> E/MethodChannel#home_widget(11269): Failed to handle method call
E/MethodChannel#home_widget(11269): java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
E/MethodChannel#home_widget(11269): 	at es.antonborri.home_widget.HomeWidgetPlugin.onMethodCall(HomeWidgetPlugin.kt:99)
E/MethodChannel#home_widget(11269): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#home_widget(11269): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/MethodChannel#home_widget(11269): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:818)
E/MethodChannel#home_widget(11269): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#home_widget(11269): 	at android.os.MessageQueue.next(MessageQueue.java:326)
E/MethodChannel#home_widget(11269): 	at android.os.Looper.loop(Looper.java:160)
E/MethodChannel#home_widget(11269): 	at android.app.ActivityThread.main(ActivityThread.java:6669)
E/MethodChannel#home_widget(11269): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#home_widget(11269): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/MethodChannel#home_widget(11269): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/flutter (11269): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: PlatformException(error, java.lang.Integer cannot be cast to java.lang.Long, null, java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Long
E/flutter (11269): 	at es.antonborri.home_widget.HomeWidgetPlugin.onMethodCall(HomeWidgetPlugin.kt:99)
E/flutter (11269): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/flutter (11269): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/flutter (11269): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:818)
E/flutter (11269): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/flutter (11269): 	at android.os.MessageQueue.next(MessageQueue.java:326)
E/flutter (11269): 	at android.os.Looper.loop(Looper.java:160)
E/flutter (11269): 	at android.app.ActivityThread.main(ActivityThread.java:6669)
E/flutter (11269): 	at java.lang.reflect.Method.invoke(Native Method)
E/flutter (11269): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/flutter (11269): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/flutter (11269): )
E/flutter (11269): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:581:7)
E/flutter (11269): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
E/flutter (11269): <asynchronous suspension>
E/flutter (11269): 

When I try clicking on a part of a home widget I get an error:

> E/HomeWidgetBackgroundService(11269): No callbackHandle saved. Did you call HomeWidgetPlugin.registerBackgroundCallback?

And then the app crashes. 

I think I found a solution for this problem. 
If I change https://github.com/ABausG/home_widget/blob/4f692310883a1bc1aeebb4c79af54b25b2bc4905/android/src/main/kotlin/es/antonborri/home_widget/HomeWidgetPlugin.kt#L95 and the next line to:
```
val dispatcher = (call.arguments as Iterable<*>).toList()[0].toString().toLong()

val callback = (call.arguments as Iterable<*>).toList()[1].toString().toLong()
```
The issue with casting is gone and app registers a background callback and everything works as expected.

Best regards,
Aljosa

**Comment by ABausG** [comment]
Thanks for the report. Just to be safe could you also show the code section where you call `HomeWidget.registerBackgroundCallback` and even more important the full method signature of the callback you try to invoke? It needs to be a static function

**Comment by aljkor** [comment]
```
/// Called when Doing Background Work initiated from Widget
void backgroundCallback(Uri data) async {
  print(data);
  print(data.host);

  if (data.host == 'titleclicked') {
    final greetings = [
      'Hello',
      'Hallo',
      'Bonjour',
      'Hola',
      'Ciao',
      '哈洛',
      '안녕하세요',
      'xin chào'
    ];
    final selectedGreeting = greetings[Random().nextInt(greetings.length)];

    await HomeWidget.saveWidgetData<String>('title', selectedGreeting);
    await HomeWidget.updateWidget(
        name: 'HomeWidgetMyProvider', iOSName: 'HomeWidgetExample');
  } else if (data.host == 'refreshclicked'){
    print('refreshclicked');
    await HomeWidget.saveWidgetData<int>('pollenIndex', Random().nextInt(10));
    await HomeWidget.saveWidgetData<int>('pollutantIndex', Random().nextInt(10));
    await HomeWidget.updateWidget(
        name: 'HomeWidgetMyProvider', iOSName: 'HomeWidgetExample');
  }
}


class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
    HomeWidget.setAppGroupId('YOUR_GROUP_ID');
    HomeWidget.registerBackgroundCallback(backgroundCallback);
  }
```

**Comment by ABausG** [comment]
Ah so this is for the example app? What device/Android Version did you encounter this issue?

**Comment by aljkor** [comment]
Not exactly example app but it is largely based on the example app yes.
I tested on emulators in Android Studio (Nexus 6 API 28 and Pixel 3a API 30).
After I changed those two lines to `.toString().toLong()` it worked as expected. I have a button that when clicked updates two progress bars with random progress in home widget. Before I changed those two lines app crashed when I clicked refresh, after I did it I get the expected result of progress bars randomly changing values when I click refresh button.
